File: leetcode/455.cpp

// Explanation:
// first sort the two array
// then use two pointer to traverse the two array
// if the value of s[index] >= g[cnt] then cnt++
// else index++
// return cnt
// Time complexity: O(nlogn)
// Space complexity: O(1)

class Solution {
public:
  int findContentChildren(vector<int> &g, vector<int> &s) {
    if (g.size() == 0 || s.size() == 0)
      return 0;
    sort(g.begin(), g.end());
    sort(s.begin(), s.end());
    int index = 0, cnt = 0;
    while (index < s.size() && cnt < g.size()) {
      if (s[index] >= g[cnt])
        cnt++;
      index++;
    }
    return cnt;
  }
};

File: leetcode/455.rs
// Explanation
// Sort the children and cookies in ascending order.
// For each cookie, check if it can satisfy the greed of the child.
// If it can, increment the count and move to the next child.
// Use count as the index of the greed vector.
// If the cookie is not large enough, move to the next cookie.
// if there are no more cookies, break out of the loop.

use std::convert::TryInto;
impl Solution {
    pub fn find_content_children(mut g: Vec<i32>, mut s: Vec<i32>) -> i32 {
        g.sort_unstable();
        s.sort_unstable();
        let mut count = 0;

        for cookie in &s {
            if let Some(greed) = g.get(count) {
                if cookie >= greed {
                    count += 1;
                }
            } else {
                break; // No more children to satisfy
            }
        }

        count.try_into().unwrap()
    }
}
